lib/Settings/Settings.cpp

```diff
@@ -0,0 +1,223 @@
+/**
+  MQTT433gateway - MQTT 433.92 MHz radio gateway utilizing ESPiLight
+  Project home: https://github.com/puuu/MQTT433gateway/
+
+  The MIT License (MIT)
+
+  Copyright (c) 2017 Jan Losinski
+
+  Permission is hereby granted, free of charge, to any person
+  obtaining a copy of this software and associated documentation files
+  (the "Software"), to deal in the Software without restriction,
+  including without limitation the rights to use, copy, modify, merge,
+  publish, distribute, sublicense, and/or sell copies of the Software,
+  and to permit persons to whom the Software is furnished to do so,
+  subject to the following conditions:
+
+  The above copyright notice and this permission notice shall be
+  included in all copies or substantial portions of the Software.
+
+  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+  EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+  MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+  NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+  BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+  ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+  CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+  SOFTWARE.
+*/
+#include <algorithm>
+
+#include "Settings.h"
+
+#include <ArduinoJson.h>
+#include <FS.h>
+
+#include <ArduinoSimpleLogging.h>
+
+static inline bool any(std::initializer_list<bool> items) {
+  return std::any_of(items.begin(), items.end(),
+                     [](bool item) { return item; });
+}
+
+std::function<bool(const String &)> notEmpty() {
+  return [](const String &str) { return str.length() > 0; };
+}
+
+template <typename T>
+std::function<bool(const T &)> notZero() {
+  return [](const T &val) { return val != 0; };
+}
+
+template <typename TVal, typename TKey>
+bool setIfPresent(JsonObject &obj, TKey key, TVal &var,
+                  const std::function<bool(const TVal &)> &validator =
+                      std::function<bool(const TVal &)>()) {
+  if (obj.containsKey(key)) {
+    TVal tmp = obj.get<TVal>(key);
+    if (tmp != var && (!validator || validator(tmp))) {
+      var = tmp;
+      return true;
+    }
+  }
+  return false;
+}
+
+void Settings::registerChangeHandler(SettingType setting,
+                                     const SettingCallbackFn &callback) {
+  listeners.emplace_front(setting, callback);
+}
+
+void Settings::onConfigChange(SettingTypeSet typeSet) const {
+  for (const auto &listener : listeners) {
+    if (typeSet[listener.type]) {
+      listener.callback(*this);
+    }
+  }
+}
+
+void Settings::load() {
+  if (SPIFFS.exists(SETTINGS_FILE)) {
+    File file = SPIFFS.open(SETTINGS_FILE, "r");
+    String settingsContents = file.readStringUntil(SETTINGS_TERMINATOR);
+    file.close();
+    Logger.debug.println("FILE CONTENTS:");
+    Logger.debug.println(settingsContents);
+
+    deserialize(settingsContents, false);
```

Comment:
`deserialize()` should directly read from Stream. Please provide a overloaded function. Then `SETTINGS_TERMINATOR` is no longer necessary.